boilermaker/app.py

```diff
@@ -263,15 +282,19 @@ async def message_handler(
                 await self.publish_task(task.on_success)
         except RetryException as retry:
             # A retry has been requested:
-            # no on_failure run until after retries exhausted
+            # Calculate next delay and publish retry.
+            # Do not run on_failure run until after retries exhausted!
+            # The `retry` RetryException may have a policy -> What if it's different from the Task?
+            if retry.policy and retry.policy != task.policy:
+                # This will publish the *next* instance of the task using *this* policy
+                task.policy = retry.policy
+                logger.warning(f"Task policy updated to retry policy {retry.policy}")
```

Comment:
This is surprising usage. I can't imagine a scenario where in which a variable retry like this would be necessary. Was there a use case that motivated this? This feels like a footgun waiting to happen but I might be missing something.

Reply:
This is specifically for scenarios like this:

```python
async def a_background_task(state, param: str):
    # In some scenarios, from within the function I want a different policy
    if param == "throwing":
        raise retries.RetryExceptionDefaultExponential("We are thrown", max_tries=42)
    return "OK"
```

The above function will use the policy associated with the exception thrown and _not_ the one it was registered with. I had originally intended this behavior and in one project I'm trying to throw with custom retries, but before this PR there's no way to respect that.

Reply:
One other note: this assignment _doesn't_ change the value in the task registry. It changes the published task to have a different policy, which should carry throw for all future retry-invocations, unless someone does something weird with dynamically shifting the policy within the same handler.

Reply:
I'm trying to wrap my head around when one would want a different retry policy than the retry policy in the original task. Do you have a concrete example where you needed it? That might help me understand this scenario better. In my mind it still feels like it would be more sensible to spawn a new task with the new retry policy rather than dynamically change it from within the existing task? 

```python
async def my_task_with_policy_1(state, param: str):
    result = do_normal_thing()
    if result.retry_state:
        raise RetryException() # Just retry according to policy for this task
    elif result.weird_state:
        # For some reason here, we need a different retry than original 
        publish_task(new_task_with_new_policy)
```

Reply:
Oh, it's possible that there are different types of errors we can hit from within a task:

```python
def some_task():
  # Do something sorta reliable, but it fails
  raise RetryException(...)
  # Do something highly reliable, but if it's failing in this case, something super weird must have occurred....
  raise RetryExceptionOfADifferentKind(...)
```

The task has knowledge and may wish to adjust how it's responding to conditions around it.

Reply:
okay, I'm going to merge for now: we can revisit how Retries work in the future. We may want to stop using exceptions for control-flow at that time...